login-module/src/core.js

```diff
@@ -0,0 +1,77 @@
+import constants from './constants'
+import { WebAuth } from 'auth0-js'
+import Quiz from './quiz'
+import Enrollment from './enrollment'
+import Certificate from './certificate'
+
+export default class GraphAcademyCore {
+	constructor(options = {}) {
+		this.options = { ...constants.DEFAULT_OPTIONS, ...options }
+		const hasRequiredOptions = this.hasRequiredOptions(options)
+		if (!hasRequiredOptions) {
+			console.log(`required params missing - one of ${constants.REQUIRED_OPTIONS.join(', ')}`)
+			return
+		}
+		this.webAuth = new WebAuth({
+			clientID: 'hoNo6B00ckfAoFVzPTqzgBIJHFHDnHYu',
+			domain: 'login.neo4j.com',
+			redirectUri: `${window.location.origin}/accounts/login`,
+			audience: 'neo4j://accountinfo/',
+			scope: 'read:account-info openid email profile user_metadata',
+			responseType: 'token id_token'
+		})
+		this.certificate = new Certificate(this.options.trainingClassName, this.options.stage)
+		this.enrollment = new Enrollment(this.options.trainingClassName, this.options.stage)
+		this.quiz = new Quiz(this.options.trainingClassName, this.options.stage)
+	}
+
+	hasRequiredOptions(options) {
+		return constants.REQUIRED_OPTIONS.every(item => options[item])
+	}
+
+	async checkSession() {
+		const { options, webAuth } = this
+		return new Promise((resolve, reject) => {
+			webAuth.checkSession({}, async (err, result) => {
+				if (err) {
+					delete this.authResult
+					if (options.loginRedirectUrl) {
+						this.redirectToLogin()
+					}
+					return reject(err)
+				}
+				this.authResult = result
+				resolve(result)
+			})
+		})
+	}
+
+	async enrollStudentInClass (firstName, lastName) {
+		return await this.enrollment.enrollStudentInClass(firstName, lastName, await this.getAccessToken())
```

Comment:
Let's not have `await` in the fn param. Move it above this and then simply pass it a regular param. But additionally, can we try to have `getAccessToken` as a non-async function? I believe that `checkSession` already sets `result` as a class property. So when you call `getAccessToken` it is always available, correct?

Reply:
> Let's not have await in the fn param. Move it above this and then simply pass it a regular param.

:+1: 

>  But additionally, can we try to have getAccessToken as a non-async function?

`getAccessToken` will call `checkSession` (async) if you don't already have an authentification context (`authResult`) or return the access token from `authResult` (instantaneous).

>  I believe that checkSession already sets result as a class property. So when you call getAccessToken it is always available, correct?

Not really, consider the following code:

```js
const graphAcademy = new GraphAcademy({})
graphAcademy.enrollStudentInClass('Guillaume', 'Grossetie') // accessToken will be undefined
```

I don't think we should call `checkSession` in the `constructor` because, in my opinion, a constructor should not do IO/async calls. With `getAccessToken` you are guaranteed to get an access token (or an error).
We could also have an accessor on `authResult` if you want to get the current authentification context but this value can be undefined/null:

```js
const graphAcademy = new GraphAcademy({})
graphAcademy.getAuthResult() // null/undefined

await graphAcademy.checkSession()
graphAcademy.getAuthResult() // { acessToken: '...' }
```

Reply:
So we should definitely have `checkSession` called explicitly and not do that as a part of the constructor. But we should also assume/force that we call it before `getAccessToken()` or `getAuthResult()`. If these functions are called before `checkSession`, they would simply return null. And then we would not need to have these getters as async functions.

Reply:
>  If these functions are called before checkSession, they would simply return null. And then we would not need to have these getters as async functions.

I want to avoid that using an "authenticated object" as described above. Let me know if it makes sense.